File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int seconds = 0;
        int[] bridge = new int[bridge_length];
        
        Deque<Integer> deque = new ArrayDeque<>();
        for(int truck : truck_weights) {
            deque.addLast(truck);
        }
        
        while(!deque.isEmpty() || checkBridgeWeight(bridge) > 0) {
            seconds++;
            moveTruck(bridge);
            if(!deque.isEmpty() && checkBridgeWeight(bridge) + deque.peekFirst() <= weight) {
                bridge[0] = deque.pollFirst();
            }
        }
        
        return seconds;
    }
    
    public int checkBridgeWeight(int[] bridge) {
        int weight = 0;
        
        for(int w : bridge) {
            weight += w;
        }
        
        return weight;
    }
    
    public void moveTruck(int[] bridge) {
        for(int i = bridge.length - 1; i > 0; i--) {
             bridge[i] = bridge[i - 1];
        }
        
        bridge[0] = 0;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    time = 0
    current_weight = 0
    bridge_queue = deque([0] * bridge_length)
    truck_weights = deque(truck_weights)
    
    while truck_weights:
        current_weight -= bridge_queue.popleft()
        
        if truck_weights[0] + current_weight <= weight:
            truck = truck_weights.popleft()
            current_weight += truck
            bridge_queue.append(truck)
            
        else:
            bridge_queue.append(0)
            
        time += 1
        
    return time + bridge_length